doghouse/appengine/github.go

```diff
@@ -0,0 +1,236 @@
+	if err != nil {
+		log.Errorf(ctx, "failed to get access token: %v", err)
+		w.WriteHeader(http.StatusBadRequest)
+		fmt.Fprintln(w, "failed to get GitHub access token")
+		return
+	}
+	g.tokenStore.Set(w, []byte(token))
+
+	// Redirect.
+	redirURL := "/gh/"
+	if r, _ := g.redirURLStore.Get(r); err == nil {
+		redirURL = string(r)
+		g.redirURLStore.Clear(w)
+	}
+	http.Redirect(w, r, redirURL, http.StatusFound)
+}
+
+func (g *GitHubHandler) Handler(h http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		ctx := appengine.NewContext(r)
+		if g.isLoggedIn(r) {
+			h.ServeHTTP(w, r)
+			return
+		}
+		// Not logged in yet.
+		log.Debugf(ctx, "Not logged in yet.")
+		state := securerandom(16)
+		g.redirURLStore.Set(w, []byte(r.URL.RequestURI()))
+		g.authStateStore.Set(w, []byte(state))
+		http.Redirect(w, r, g.buildGithubAuthURL(r, state), http.StatusFound)
+	})
+}
+
+func (g *GitHubHandler) isLoggedIn(r *http.Request) bool {
+	ok, _ := g.token(r)
+	return ok
+}
+
+func securerandom(n int) string {
+	state := make([]byte, n)
+	io.ReadFull(rand.Reader, state[:])
+	return fmt.Sprintf("%x", state)
+}
+
+// https://developer.github.com/apps/building-github-apps/identifying-and-authorizing-users-for-github-apps/#2-users-are-redirected-back-to-your-site-by-github
+// POST https://github.com/login/oauth/access_token
+func (g *GitHubHandler) requestAccessToken(ctx context.Context, code, state string) (string, error) {
+	const u = "https://github.com/login/oauth/access_token"
+	cli := urlfetch.Client(ctx)
+	data := url.Values{}
+	data.Set("client_id", g.clientID)
+	data.Set("client_secret", g.clientSecret)
+	data.Set("code", code)
+	data.Set("state", state)
+
+	req, err := http.NewRequest("POST", u, strings.NewReader(data.Encode()))
+	if err != nil {
+		return "", fmt.Errorf("failed to create request: %v", err)
+	}
+	req = req.WithContext(ctx)
+	req.Header.Add("Accept", "application/json")
+	req.Header.Add("Accept", "application/vnd.github.machine-man-preview+json")
+	req.Header.Add("Content-Type", "application/x-www-form-urlencoded")
+
+	res, err := cli.Do(req)
+	if err != nil {
+		return "", fmt.Errorf("failed to request access token: %v", err)
+	}
+	defer res.Body.Close()
+
+	b, _ := ioutil.ReadAll(res.Body)
+
+	var token struct {
+		AccessToken string `json:"access_token"`
+	}
+	if err := json.NewDecoder(bytes.NewReader(b)).Decode(&token); err != nil {
+		return "", fmt.Errorf("failed to decode response: %v", err)
+	}
+
+	if token.AccessToken == "" {
+		log.Errorf(ctx, "response doesn't contain token (resopnse: %s)", b)
+		return "", errors.New("response doesn't contain GitHub access token")
+	}
+
+	return token.AccessToken, nil
+}
+
+func (g *GitHubHandler) token(r *http.Request) (bool, string) {
+	b, err := g.tokenStore.Get(r)
+	if err != nil {
+		return false, ""
+	}
+	return true, string(b)
+}
+
+func (g *GitHubHandler) HandleGitHubTop(w http.ResponseWriter, r *http.Request) {
```

Comment:
**[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
exported method GitHubHandler.HandleGitHubTop should have comment or be unexported